File: cs/ch17-04-arithmetic/Money/Money.cs
namespace TheSoftwareGorilla.TDD.Money;

public class Money : ICurrencyHolder<Money>, IExpression<Money>
{

    public static Money From(Decimal amount, String currency, Bank<Money> bank)
    {
        return new Money(amount, currency, bank);
    }

    public decimal Amount { get; }

    public string Currency { get; }

    public Bank<Money> Bank { get; }

    protected Money(decimal amount, string currency, Bank<Money> bank)
    {
        Amount = Math.Round(amount, 2, MidpointRounding.AwayFromZero);
        Currency = currency;
        Bank = bank;
    }

    public virtual Money Add(Money addend)
    {
        Func<Money, Money, Money> plus = (aue, ade) => Money.From(aue.Amount + ade.Amount, aue.Currency, aue.Bank);
        return InvokeOperator(addend, plus);
    }

    public static Money operator +(Money augend, Money addend)
    {
        return augend.Add(addend);
    }

    public virtual Money Subtract(Money subtrahend)
    {
        Func<Money, Money, Money> minus = (minuend, subtrahend) => Money.From(minuend.Amount - subtrahend.Amount, minuend.Currency, minuend.Bank);   
        return InvokeOperator(subtrahend, minus);
    }
    
    public static Money operator -(Money minuend, Money subtrahend)
    {
        return minuend.Subtract(subtrahend);
    }

    public virtual Money Divide(decimal divisor)
    {
        return Money.From(Amount / divisor, Currency, Bank);
    }

    public static Money operator /(Money dividend, decimal divisor)
    {
        return dividend.Divide(divisor);
    }

    public virtual Money Multiply(decimal multiplier)
    {
        return Money.From(Amount * multiplier, Currency, Bank);
    }

    public static Money operator *(Money multiplicand, decimal multiplier)
    {
        return multiplicand.Multiply(multiplier);
    }


    protected virtual Money InvokeOperator(Money other, Func<Money, Money, Money> operation)
    {
        Operation<Money> operationInstance = new Operation<Money>(this, other, Bank, Currency, operation);
        return operationInstance.Apply();
    }

    public virtual Money Convert(string to)
    {
        return Bank.Convert(this, to);
    }


    public override bool Equals(object? obj)
    {
        return obj is Money money 
            && Amount == money.Amount 
            && Currency == money.Currency;
    }

    public static bool operator ==(Money left, Money right)
    {
        return left.Equals(right);
    }

    public static bool operator !=(Money left, Money right)
    {
        return !(left == right);
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Amount, Currency);
    }

    public override string ToString()
    {
        return Amount + " " + Currency;
    }

    public decimal AmountIn(string currency)
    {
        return Convert(currency).Amount;
    }

    public virtual Money NewCurrencyHolder(decimal amount, string currency, Bank<Money> bank)
    {
        return Money.From(amount, currency, bank);
    }
}

File: java/ch17-04-arithmetic/src/main/java/com/thesoftwaregorilla/tdd/money/Money.java
package com.thesoftwaregorilla.tdd.money;


import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.function.BiFunction;


public class Money implements ICurrencyHolder<Money>, IExpression<Money> {


    public static Money from(BigDecimal amount, String currency, Bank<Money> bank) {
        return new Money(amount, currency, bank);
    }

    private final BigDecimal amount;
    private final String currency;
    private final Bank<Money> bank;

    protected Money(BigDecimal amount, String currency, Bank<Money> bank) {
        this.amount = amount.setScale(2, RoundingMode.HALF_UP);
        this.currency = currency;
        this.bank = bank;
    }


    @Override
    public Bank<Money> getBank() {
        return bank;
    }

    @Override
    public Money convert(String toCurrency) {
        return bank.convert(this, toCurrency);
    }

    @Override
    public BigDecimal getAmount() {
        return amount;
    }

    @Override
    public BigDecimal getAmountIn(String toCurrency) {
        return convert(toCurrency).getAmount();
    }

    @Override
    public String getCurrency() {
        return currency;
    }

    @Override
    public Money newCurrencyHolder(BigDecimal amount, String currency, Bank<Money> bank) {
        return Money.from(amount, currency, bank);
    }

    @Override
    public Money add(Money addend) {
        BiFunction<Money, Money, Money> operation = (a, b) -> Money.from(a.getAmount().add(b.getAmount()), a.getCurrency(), a.getBank());
        return InvokeOperator(addend, operation);
    }

    @Override
    public Money subtract(Money subtrahend) {
        BiFunction<Money, Money, Money> operation = (a, b) -> Money.from(a.getAmount().subtract(b.getAmount()), a.getCurrency(), a.getBank());
        return InvokeOperator(subtrahend, operation);
    }

    @Override
    public Money multiply(BigDecimal multiplier) {
        return Money.from(getAmount().multiply(multiplier), getCurrency(), getBank());
    }

    @Override
    public Money divide(BigDecimal divisor) {
        return Money.from(getAmount().divide(divisor, 2, RoundingMode.HALF_UP), getCurrency(), getBank());
    }

    protected Money InvokeOperator(Money other, BiFunction<Money, Money, Money> operation) {
        Operation<Money> operationInstance = new Operation<>(this, other, bank, currency, operation);
        return operationInstance.evaluate();
    }

    @Override
    public boolean equals(Object object) {
        return object instanceof ICurrencyHolder<?> money
                && getAmount().equals(money.getAmount())
                && getCurrency().equals(money.getCurrency());
    }

    @Override
    public int hashCode() {
        return getAmount().hashCode() + getCurrency().hashCode() + getBank().hashCode();
    }

    @Override
    public String toString() {
        return amount + " " + currency;
    }
}
